NIFI-15932: Implemented ability to migrate a Versioned flow's Assets … (#11240)
apache/nifi: nifi-docs/src/main/asciidoc/administration-guide.adoc

|`nifi.secrets.manager.cache.duration`|The duration for which resolved secret values are cached before being refreshed from the underlying Parameter Providers. Accepts any NiFi time duration value such as `5 mins`, `30 secs`, etc. A value of `0 sec` disables caching entirely. Defaults to `5 mins`.
|====

== Migrating a Connector from a Versioned Process Group

NiFi supports a one-time path that migrates a newly created Connector from a source Versioned Process Group.
The source Versioned Process Group is used as a reference: the Connector reads it and updates its own managed flow to mirror the source's configuration, parameters, and component state.
The source flow itself is not installed onto the Connector.

> Only Connectors that explicitly opt in to migration support can use this capability.

Reviewer: ```suggestion
Only Connectors that implement migration support can use this capability.
```

They opt in by implementing the support interfaces. While this is outlined in the developer guide an administrator may confuse opt in for some type of configuration rather than code.

Author: [Sonnet 5] Updated as suggested.

=== Prerequisites for a local migration

When the source is a local Process Group on the same NiFi instance, the Process Group must satisfy all of the following requirements before NiFi will allow the migration:

- The source Process Group must be under version control.
- The source Process Group must be `UP_TO_DATE` with the latest published version in Flow Registry.
- All processors in the source Process Group must be stopped.
- All controller services in the source Process Group must be disabled.
- All queues in the source Process Group must be empty.
- The source Process Group must not reference controller services outside of the Process Group.

> - The target Connector must be newly created, stopped, and not previously migrated.

Reviewer: ```suggestion
- The target Connector must be not yet configured, stopped, and not previously migrated.
```
How recent the connector was created does not seem to come into play. It's whether it was configured prior to the attempt to migrate.

Author: [Sonnet 5] Updated as suggested.

- The target Connector's active flow must be empty when migration begins. Connectors that ship a non-empty initial flow are not compatible with the migration path.

If the source Process Group is not under version control, export it and use the uploaded payload path instead.

=== Migration paths

There are two ways to run the migration:

- Local Process Group migration:
  Select an eligible Process Group that already exists on the current NiFi canvas.
  NiFi validates the prerequisites listed above before starting the migration.
- Uploaded payload migration:

>   Export a Process Group definition and upload the resulting flow snapshot payload to the Connector migration endpoint.

Reviewer: May want to mention the exports should include state (if it exists).

Author: [Sonnet 5] Good call — updated the uploaded payload migration step to mention exporting component state.

  This path is intended for non-version-controlled sources, sources exported from another NiFi instance, or any other source that cannot satisfy the local version-control requirements.

For uploaded payloads, NiFi does not require the source flow to be under version control.
The Connector's own migration support check determines whether the uploaded flow can be imported.

=== Running a migration

. Create a new Connector instance.
. If the source Process Group is local and registry-backed, choose it from the Connector migration source list.
. If the source Process Group is not eligible for the local path, export it with component state included and upload the payload to the Connector migration payload endpoint.
. Start the Connector migration request.
. Poll the request until it completes.

> . Open the Connector configuration and supply any missing sensitive values or secrets.

Reviewer: ```suggestion
. Open the Connector configuration and supply any missing secrets.
```

Since only Connector Secrets are supported I'm not sure mentioning sensitive values make sense unless referring to the source flow.

Author: [Sonnet 5] Updated as suggested.
